raiden_contracts/contracts/MonitoringService.sol

```diff
@@ -0,0 +1,321 @@
+pragma solidity ^0.4.23;
+
+import "./Token.sol";
+import "./Utils.sol";
+import "./lib/ECVerify.sol";
+import "./TokenNetwork.sol";
+
+contract MonitoringService is Utils {
+
+    // Token to be used for paying the rewards
+    Token public token; // Only allow RDN?
+
+    // The minimum allowed deposit for a MS
+    uint256 public minimum_deposit;
+
+    // keep track of registered Monitoring Services
+    // A Monitoring Services must deposit to register
+    mapping(address => bool) public registered_monitoring_services;
+
+    // channel_identifier => Struct
+    // Keep track of the rewards per channel
+    mapping(bytes32 => Reward) rewards;
+
+    // Keep track of balances
+    mapping(address => uint) public balances;
+
+    /*
+     *  Structs
+     */
+    struct Reward{
+        // The signature of the reward
+        bytes reward_proof_signature;
+
+        // Nonce of the most recently provided BP
+        uint256 nonce;
+
+        // Address of the Raiden Node that was monitored
+        // This is also the address that has the reward deducted from its deposit
+        address reward_sender_address;
+    }
+
+    /*
+     *  Events
+     */
+
+    //event MonitoringServiceNewDeposit(address sender, address receiver, uint amount);
+    event NewDeposit(address sender, address receiver, uint amount);
+    event RegisteredMonitoringService(address monitoring_service);
+    event NewBalanceProofReceived(
+        bytes reward_proof_signature,
+        uint256 nonce,
+        address ms_address,
+        address raiden_node_address
+    );
+    event RewardClaimed(address ms_address, uint amount, bytes32 channel_id);
+    event MonitoringServiceRegistered(address ms_address);
+    event MonitoringServiceDeregistered(address ms_address);
+    event Withdrawn(address account, uint amount);
+
+    /*
+     *  Modifiers 
+     */
+
+    modifier isMonitor(address _ms_address) {
+        require(registered_monitoring_services[_ms_address]);
+        _;
+    }
+
+    modifier isNotMonitor(address raiden_node_address) {
+        require(!registered_monitoring_services[raiden_node_address]);
+        _;
+    }
+
+    /*
+     *  Constructor
+     */
+
+    /// @notice Set the default values for the smart contract
+    /// @param _token_address The address of the token to use for rewards
+    /// @param _minimum_deposit The minimum value that a MS must deposit to be registered
+    constructor(
+        address _token_address,
+        uint256 _minimum_deposit
```

Comment:
This should be modifiable. Token value will fluctuate.
Drawbacks - adds an `owner` role with some responsibilities, so it must be discussed.

Reply:
It would really be nice to not have an owner, but it's a good point that I didn't really consider. I guess an ideal solution would include an oracle that could update the value according to the price every day, but that would also make a mess, if an MS then needs to deposit more tokens to be re-registered. We need to think about this indeed.

Reply:
my idea was to have an auction contract that would reset the registry each N blocks and the MSs would be re-elected depending on deposit they've put in.